### PR TITLE
Make kubernetes and application-engine type current contexts mutual exclusive

### LINKED ISSUE #114: Make kubernetes and application-engine type current contexts mutual exclusive
**Describe the feature request**
The application-engine and kubernetes contexts types are similar in nature because `application-engine` would also be pointing to k8s cluster. Since they both are similar in nature, it is decided to have either of the context types to be active. This would have impact on context based plugins that would be installed or uninstalled based on the current context.
**Describe alternative(s) you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by prkalle**
### What this PR does / why we need it
This PR makes kubernetes and application-engine current contexts mutual exclusive, which means if the k8s context is set as current(active) the TAE context would be removed and vice-versa.

### Which issue(s) this PR fixes

Fixes #114 

### Describe testing done for PR
Unit tests and CI passed.

### Release note

```release-note
Make kubernetes and application-engine type current contexts mutual exclusive
```

### CODE REVIEW

**Comment by prkalle** [comment]
> We need to add Compatibility tests for this change as well.

I will add the Compatibility tests in a follow up PR.

File: config/contexts_test.go [hunk 1]
```diff
@@ -1284,3 +1263,103 @@ func TestSetContextWithEmptyName(t *testing.T) {
...
+
+	Context("tmc context as current context", func() {
+		It("should set tmc context as current context successfully", func() {
+			err = SetCurrentContext("test-tmc")
+			gomega.Expect(err).To(gomega.BeNil())
+			validateCurrentContext(configtypes.TargetTMC, "test-tmc", true, "test-mc-2")
+		})
+	})
+	Context("k8s context as current context", func() {
+		It("should set tmc context as current context successfully", func() {
```
Comment (marckhouzam) [inline config/contexts_test.go:1290]: title is wrong

**Review by marckhouzam — APPROVED** [review]
Thanks @prkalle !
Two nits, but LGTM

File: config/contexts.go [hunk 2]
```diff
@@ -443,3 +449,33 @@ func removeContext(node *yaml.Node, name string) error {
 	contextsNode.Content = contexts
 	return nil
 }
+
+// updateMutualExclusiveCurrentContexts updates the current contexts to maintain
+// mutual exclusive behavior between K8S and TAE current context types
+func updateMutualExclusiveCurrentContexts(node *yaml.Node, ctx *configtypes.Context) error {
```
Comment (vuil) [inline config/contexts.go:455]: nit: (I don't feel super strongly about this) I am wondering if we can write this in a more future proof manner by checking for TargetTMC instead.
iow, 
when ctx is not tmc:
   deactivate any ctx that is not tmc

then the code works even when new context types are introduced

File: config/contexts_test.go [hunk 3]
```diff
@@ -6,13 +6,21 @@ package config
 import (
 	"testing"
 
+	. "github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
```
[the remark this answers is not in the document]
Comment (vuil) [reply config/contexts_test.go:10]: +1

File: config/contexts_test.go [hunk 4]
```diff
@@ -1284,3 +1263,103 @@ func TestSetContextWithEmptyName(t *testing.T) {
...
+			validateCurrentContext(configtypes.TargetTMC, "test-tmc", true, "test-mc-2")
+
+			err = SetCurrentContext("test-mc")
+			gomega.Expect(err).To(gomega.BeNil())
+			validateCurrentContext(configtypes.TargetK8s, "test-mc", true, "test-mc")
+
+		})
+	})
+	Context("k8s context as current context after tae context(mutual-exclusion test between k8s and tae) ", func() {
+		It("should have only k8s as current context and tae context should be removed from the curren context", func() {
```
Comment (vuil) [inline config/contexts_test.go:1322]: typo curren, same for next It as well.

File: config/contexts_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (prkalle) [reply config/contexts_test.go:10]: It was not imported into the current package on purpose as the gomega has global variable `Default` which was conflicting with `Default` global variable declared in `config` package.

File: config/contexts_test.go (hunk 1 above)
Comment (prkalle) [reply config/contexts_test.go:1290]: Thanks, fixed.

File: config/contexts_test.go (hunk 4 above)
Comment (prkalle) [reply config/contexts_test.go:1322]: fixed.

**Review by marckhouzam — COMMENTED** [review]
@prkalle There is another `SetCurrentContext()` function in `clientconfig.go`.  It is not clear to me when that one is used, but I get the feeling it can and therefore it may need to be updated as well?

**Review by marckhouzam — CHANGES_REQUESTED** [review]
Blocking the PR until we discuss the last comment https://github.com/vmware-tanzu/tanzu-plugin-runtime/pull/113#pullrequestreview-1664937527

I was just afraid the comment would be missed since I added it after approving

File: config/contexts.go (hunk 2 above)
Comment (prkalle) [reply config/contexts.go:457]: Sure. Updated the PR to maintain mutual exclusion behavior among all the current context types except TMC. Thanks

**Comment by prkalle** [comment]
> @prkalle There is another `SetCurrentContext()` function in `clientconfig.go`. It is not clear to me when that one is used, but I get the feeling it can and therefore it may need to be updated as well?

`SetCurrentContext()` function in `clientconfig.go` is just to modify the current contexts in the client and user has to call `StoreClientConfig()` method inorder to persist the changes to the config file, which would internally call  `clientConfigSetCurrentContext()` that would endup calling the `setCurrentContext()` in context.go file. So we should be good without updating the `SetCurrentContext()` function in `clientconfig.go`.

cc: @mpanchajanya Thanks for clarification

**Comment by marckhouzam** [comment]
Thanks for the explanation @prkalle (and @mpanchajanya ).  It makes sense.

**Review by marckhouzam — APPROVED** [review]
LGTM
Thanks @prkalle !

**Pull request merged**